### ISSUE TITLE
[FEATURE] Covert LatLng to Point with map rotation in mind

### ISSUE THREAD

**Issue by mohammedX6** [fleaflet/flutter_map#1311]
### What do you want implemented?

I want to covert a LatLng to point with taking care map rotation

minimal example of the problem I have https://filebin.net/t8uftfm4aklz69yx/minimal_example.zip

### What other alternatives are available?

I tried this but the result is wrong 
```
 @override
  CustomPoint? latLngToPoint(LatLng localPoint) {
    if (_state.originalSize == null) {
      return null;
    }

    final width = _state.originalSize!.x;
    final height = _state.originalSize!.y;

    var pos = _state.options.crs.latLngToPoint(localPoint, _state.zoom);

    final localPointCenterDistance =
        CustomPoint((width / 2) - pos.x, (height / 2) - pos.y);

    var mapCenter =
        _state.options.crs.latLngToPoint(_state.center, _state.zoom);
    var point = mapCenter - localPointCenterDistance;

    if (_state.rotation != 0.0) {
      point = point - _state.getPixelOrigin();
      point = rotatePoint(mapCenter, point);
    }


    return point;
  }

```

### Platforms Affected

Android, iOS

### Severity

Obtrusive: No workarounds are available, and this is essential to me

### Requirements

- [X] I agree to follow this project's [Code of Conduct](https://github.com/fleaflet/flutter_map/blob/master/CODE_OF_CONDUCT.md)
- [X] I am using the [latest stable version](https://pub.dev/packages/flutter_map) of this package
- [X] I have checked for similar feature requests which may be duplicates

**Comment by ibrierley** [comment]
Whats your test case you are trying to do ?

**Comment by mohammedX6** [comment]
> Whats your test case you are trying to do ?

Hi
I am trying to create a magnifier for the map, The magnifier needs the tap position, everything is ok but if the map is rotated the tap position is wrong.

[Take a look here](https://www.flutterclutter.dev/tags/magnifier/)

**Comment by JaffaKetchup** [comment]
If you just need tap position, you can get it from `MapOptions(onTap: () {})`.

**Comment by JaffaKetchup** [comment]
Hi @mohammedX6, I have a better understanding now.

The `onTap` callback within `MapOptions` gives a `LatLng` point, but also a `TapPosition`, which provides the actual screen position, both relative to the widget (`relative`), or relative to the screen (`global`). It should work regardless of map rotation.

For example, use:
```dart
MapOptions(
  onTap: (screenPoint, _) {
    print(screenPoint.relative);
    print(screenPoint.global);
  }
),
```

Let me know if this works for you!

**Comment by JaffaKetchup** [comment]
Maybe? There might be some weird anchoring going on, if you're using markers.

**Comment by mohammedX6** [comment]
no marker involved.

**Comment by JaffaKetchup** [comment]
Where are the red circle and blue circles supposed to be?

**Comment by mohammedX6** [comment]
the red circle must be in the same location as the blue circle.
The blue circle is the point that I am currently dragging.

**Comment by mohammedX6** [comment]
I think I might dig more into the code, Anyway thanks @JaffaKetchup .

**Comment by mohammedX6** [comment]
@ibrierley Hi, here is an issue I opened, 
Anyone can use this minmial example to understand the problem I have 
https://filebin.net/t8uftfm4aklz69yx/minimal_example.zip

**Comment by ibrierley** [comment]
ok, that isn't a minimal example. But in the meantime, WHY do you want latLngToPoint. Just checking you're not trying to reinvent the wheel. (I'm not suggesting at some point there shouldn't be a method for this..but trying to understand why firstly).

For example, if you have a LatLng, why not just place a Marker there with any widget you want ? Does it need to be outside of FlutterMap or something ?

**Comment by mohammedX6** [comment]
@ibrierley YES that's exactly what i want.

**Comment by ibrierley** [comment]
I suspect this isn't easily going to happen unless someone has a lot of time and/or been working on this already. I think there's probably too many variables that one can't be confident of and change when a rotation happens. 

Maybe someone will have a brainwave though!

There's also the question of what space one wants the result in...in map widget space, in an outer widget space, or screen resolution space. 

So my interim suggestion would be either a) don't allow rotation in the app, or b) make sure everything is done in flutter_map layer coordinate space (eg as a marker widget, I would suspect this is easiest), or c) spend a lot of time hacking something together.

**Comment by ibrierley** [comment]
Actually, this may be possible if we reverse the rotationPoint...and use a pre rotated pixel origin....something like...

```
@override
  CustomPoint latLngToScreenPoint(LatLng latLng) {

    var pxo = (_state.project(_state.getCenter(),zoom) - _state.originalSize! / 2.0).round();

    var point = _state.options.crs.latLngToPoint(latLng, _state.zoom);

    final mapCenter =
    _state.options.crs.latLngToPoint(_state.center, _state.zoom) ;

    if (_state.rotation != 0.0) {
      point = rotatePoint2(mapCenter, point);
    }

    return  point - pxo;
  }

CustomPoint<num> rotatePoint2(
      CustomPoint<num> mapCenter, CustomPoint<num> point) {
    final m = Matrix4.identity()
      ..translate(mapCenter.x.toDouble(), mapCenter.y.toDouble())
      ..rotateZ(_state.rotationRad)
      ..translate(-mapCenter.x.toDouble(), -mapCenter.y.toDouble());

    final tp = MatrixUtils.transformPoint(
        m, Offset(point.x.toDouble(), point.y.toDouble()));

    return CustomPoint(tp.dx, tp.dy);
  }


```

Maybe a little messy, but I've run out of time though for today.

**Comment by mohammedX6** [comment]
MAN, THIS THING IS WORKING LIKE A CHARM !,
Where I can give a small gift like a donation or something?

**Comment by ibrierley** [comment]
Thanks for the feedback, I'll try and get a PR sorted this weekend for it (maybe with renaming the rotate functions so they're clearer), so it ends up in the main code. We don't have a Flutter_map donation thing yet (maybe that could be a discussion, not sure if we want to stay away from that or not). Normally devs will have a donation button on their respective repos though on Git if people feel flush tho ;).

**Comment by JaffaKetchup** [comment]
We probably should setup an overall donation thing.
I've already got a GitHub sponsors verified account now, and if you got one too (not difficult), GitHub allows us to add a button to this repo, and both of our names would be listed separately. This means people can donate to one or both of us freely.
I have this setup in my main repos if you want to have a look.

**Comment by mohammedX6** [comment]
> Thanks for the feedback, I'll try and get a PR sorted this weekend for it (maybe with renaming the rotate functions so they're clearer), so it ends up in the main code. We don't have a Flutter_map donation thing yet (maybe that could be a discussion, not sure if we want to stay away from that or not). Normally devs will have a donation button on their respective repos though on Git if people feel flush tho ;).
@ibrierley 
I don't see a donation button on your page, Can you send Ko-fi or PayPal donation link?

**Comment by ibrierley** [comment]
Trying to avoid spamming here, but I think you were using
https://github.com/ibrierley/flutter_map_line_editor and look halfway down.

**Comment by ibrierley** [comment]
There's now a PR at https://github.com/fleaflet/flutter_map/pull/1325 if you wanted to test that.

**Comment by mohammedX6** [comment]
I am confirmed it's working.

**Comment by JaffaKetchup** [comment]
Fixed as of v2.2.0. Thanks for the report!
